nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py

```diff
@@ -1045,6 +1043,8 @@ def build_train_valid_test_datasets(self):
         if self._test_ds is not None:
             logging.info(f'Length of test dataset: {len(self._test_ds)}')
         logging.info(f'Finished building GPT datasets.')
+        # Override limit_val_batches to be a multiple of num microbatches to prevent val_step from exiting in between a step
+        self._reconfigure_val_batches()
```

Comment:
This line was before build_valid_test_datasets before. I thought that it was meant to influence it's inner workings. So what is the purpose of it now?

Reply:
`limit_val_batches` does not impact the building of the datasets. It is just used to define the number of val batches we would like to run during every validation. So it is okay to have it after the dataset is built. The purpose behind reconfiguring `limit_val_batches` is there was a change in how limit_val_batches worked in PTL  2.0 and to accommodate that we needed the `limit_val_batches` to be a multiple of number of microbatches. The reason being in PTL < 2.0 there was a bug and limit_val_batches was not respected and we never raised a stopIteration. But with 2.0 they fixed this and StopIteration on the `dataloader_iter` gets raised when we fetch `limit_val_batches` number of microbatches. So if limit_val_batches is less than  #microbatches, this [line](https://github.com/NVIDIA/NeMo/blob/v1.20.0/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L782)  raises StopIteration leading to a hang in case of PP, if PP 1 hit a StopIteration and PP2 is waiting for the output of PP1.  So we had to add the `_reconfigure_val_batches` method to ensure `limit_val_btaches` set by user is a multiple of #microbatches and also we reconfigure in such a way that we run `limit_val_batches` number of global batches in the validation_step which is what its supposed to do.